File: test_src/fortran/hierarchical_parallelism/atomic/target__teams_distribute_parallel_for.f90

program target__teams_distribute_parallel_for
    USE OMP_LIB

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    REAL :: COUNTER = 0

    
    

     
    
    !$OMP TARGET   MAP(TOFROM: COUNTER) 



    

    
    !$OMP TEAMS DISTRIBUTE PARALLEL DO 


    DO i = 1 , L 


    

    

!$OMP ATOMIC UPDATE

counter = counter + 1.


 
     

    END DO

    !$OMP END TEAMS DISTRIBUTE PARALLEL DO
     

    !$OMP END TARGET
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target__teams_distribute_parallel_for

File: test_src/cpp/hierarchical_parallelism/atomic/float/target_teams__distribute__parallel__for.cpp
#include <iostream>
#include <limits>
#include <cmath>





bool almost_equal(float x, float y, int ulp) {

     return std::fabs(x-y) <= std::numeric_limits<float>::epsilon() * std::fabs(x+y) * ulp ||  std::fabs(x-y) < std::numeric_limits<float>::min();

}

void test_target_teams__distribute__parallel__for.cpp(){

 // Input and Outputs
 
 const int L = 5;
 const int M = 6;

float counter{};

// Main program

#pragma omp target teams  map(tofrom:counter) 

{


#pragma omp distribute 

    for (int i = 0 ; i < L ; i++ )

{


#pragma omp parallel 

{


#pragma omp for 

    for (int j = 0 ; j < M ; j++ )

{



#pragma omp atomic update

counter += float { 1 };



}

}

}

}


// Validation
if ( !almost_equal(counter,float { L*M }, 10)  ) {
    std::cerr << "Expected: " << L*M << " Got: " << counter << std::endl;
    throw std::runtime_error( "target_teams__distribute__parallel__for.cpp give incorect value when offloaded");
}

}
int main()
{
    test_target_teams__distribute__parallel__for.cpp();
}
